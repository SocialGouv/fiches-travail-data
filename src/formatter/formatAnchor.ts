import { unwrapEmail } from "../unwrapEmail";

export const formatAnchor = (node: Element): void => {
  if (node.textContent === "") {
    node.remove();
    return;
  }
  if (node.getElementsByTagName("img").length) {
    node.classList.add("no-after");
  }

  const href = node.getAttribute("href");

  // remove ATTAg(...) on pdf link
  node.removeAttribute("onclick");

  if (!href) return;

  // unwrap link with href="javascript:"
  if (/^javascript:/.test(href)) {
    node.removeAttribute("href");
    return;
  }

  if (/email-protection/.test(href)) {
    const [, data = ""] = href.split("#");
    const value = unwrapEmail(data);
    node.setAttribute("href", `mailto:${decodeURIComponent(escape(value))}`);
    return;
  }

  if (!href.match(/^\w+?:\/\//)) {
    node.setAttribute(
      "href",
      `https://travail-emploi.gouv.fr/${href.replace(/^\//, "")}`
    );
    node.setAttribute("target", "_blank");
    node.setAttribute("rel", "nofollow, noopener");
  }
};
