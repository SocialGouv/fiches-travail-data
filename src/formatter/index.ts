export { formatAnchor } from "./formatAnchor";
export { formatEmail } from "./formatEmail";
