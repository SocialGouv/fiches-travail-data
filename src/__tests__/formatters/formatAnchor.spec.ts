import { JSDOM } from "jsdom";
import { formatAnchor } from "../../formatter/formatAnchor";

const actual = `
  <body>
    <span></span>
    <img/>
    <button>foo</button>
    <button onclick="console.log('foo')">bar</button>
    <a href="">foo</a>
    <a href="relative">relative</a>
    <a href="/absolute">absolute</a>
    <a href="http://example.com">http://example.com</a>
    <a href="https://example.com">https://example.com</a>
    <a href="ftp://example.com">ftp://example.com</a>
    <a href="/foo/email-protection">bar</a>
    <a href="/foo/email-protection#">bar</a>
    <a href="/foo/email-protection#bar">qux</a>
    <a href="/foo/email-protection#73101c1d071210075e1203331200035e0306111f1a105d1501">bar</a>
    <a href="javascript">javascript</a>
    <a href="javascript:">javascript:</a>
    <a href="javascript:alert('Hello');">javascript:alert('Hello');</a>
  </body>
`;

test("formatEmail should alter the node", () => {
  const dom = new JSDOM(actual);
  const {
    document: { body },
  } = dom.window;

  Array.from(body.querySelectorAll("a")).forEach(formatAnchor);

  expect(body).toMatchSnapshot();
});
