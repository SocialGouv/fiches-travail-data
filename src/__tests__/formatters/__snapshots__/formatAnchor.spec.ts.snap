// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatEmail should alter the node 1`] = `
<a>
  foo
</a>
<a>
  bar
</a>
<a href>
  foo
</a>
<a href="https://travail-emploi.gouv.fr/relative"
   target="_blank"
   rel="nofollow, noopener"
>
  relative
</a>
<a href="https://travail-emploi.gouv.fr/absolute"
   target="_blank"
   rel="nofollow, noopener"
>
  absolute
</a>
<a href="http://example.com">
  http://example.com
</a>
<a href="https://example.com">
  https://example.com
</a>
<a href="ftp://example.com">
  ftp://example.com
</a>
<a href="mailto:">
  bar
</a>
<a href="mailto:">
  bar
</a>
<a href="mailto:">
  qux
</a>
<a href="mailto:contact-ap@asp-public.fr">
  bar
</a>
<a href="https://travail-emploi.gouv.fr/javascript"
   target="_blank"
   rel="nofollow, noopener"
>
  javascript
</a>
<a>
  javascript:
</a>
<a>
  javascript:alert('Hello');
</a>
`;
